layabout.py

```diff
@@ -1,3 +1,287 @@
+import os
+import time
+import logging
+from typing import Any, DefaultDict, Dict, Callable, List, Tuple
+from inspect import signature
+from functools import wraps
+from collections import defaultdict
+
+from slackclient import SlackClient
+
+# TODO: This is a dependency of slackclient needed for exception handling. In
+# the future, we might be able to remove this. For now we don't want to declare
+# it as an explicit dependency.
+from websocket import WebSocketConnectionClosedException
+
 __author__ = 'Reilly Tucker Siemens'
 __email__ = 'reilly@tuckersiemens.com'
-__version__ = '0.1.0'
+__version__ = '1.0.0b'
+
+# Private type alias for the complex type of the handlers defaultdict.
+_Handlers = DefaultDict[str, List[Tuple[Callable, dict]]]
+
+log = logging.getLogger(__name__)
+
+
+class LayaboutError(Exception):
+    """ Base error for all Layabout exceptions. """
+
+
+class MissingSlackToken(LayaboutError):
+    """ Raised if a Slack API token could not be found. """
+
+
+class FailedConnection(LayaboutError, ConnectionError):
+    """
+    Raised if the Slack client could not connect to the Slack API.
+
+    Inherits from both :obj:`LayaboutError` and the built-in
+    :obj:`ConnectionError` for convenience in exception handling.
+    """
+
+
+class Layabout:
+    """
+    An event handler on top of the Slack RTM API.
+
+    Args:
+        name: A unique name for this :obj:`Layabout` instance.
+        env_var: The environment variable to try to load a Slack API token
+            from.
+
+    Attributes:
+        name (str): A unique name for this :obj:`Layabout` instance.
+        slack (SlackClient): A :obj:`slackclient.SlackClient` instance.
+            Initially :obj:`None`, the attribute is set after calling the
+            :obj:`Layabout.connect` or :obj:`Layabout.run`
+            methods.
+
+    Example:
+
+        .. code-block:: python
+
+           from layabout import Layabout
+
+           layabout = Layabout('app')
+
+
+           @layabout.handle('message')
+           def echo(slack, event):
+               \"\"\" Echo all messages seen by the app. \"\"\"
+               channel = event['channel']
+               message = event['text']
+               subtype = event.get('subtype')
+
+               # Avoid an infinite loop of echoing our own messages.
+               if subtype != 'bot_message':
+                   slack.rtm_send_message(channel, message)
+
+           layabout.run()
+    """
+    def __init__(self, name: str, env_var: str = 'SLACK_API_TOKEN') -> None:
+        # TODO: Consider keyword only arguments.
+        self.name = name
+        self._env_var = env_var
+        self._token: str = None
+        self.slack: SlackClient = None
```

Comment:
My brain says this should either be assignable or a `_` attribute. Probably the inverse for _env_var.

Reply:
I originally had it as `_private`, but when I realized that I was already allowing users direct access to it via handlers
```python
@layabout.handle('hello')
def hello(slack, event):
    slack.server.websocket.close()  # Oh, snap! Now it's all broken!
```
I didn't see any reason not to make it part of the public interface. On top of that, making it part of the public interface highlighted the possibility that a user might want to use it _before_ entering the event loop to do some actions they didn't want to be doing repeatedly. This is also why I've currently made the `Layabout.connect` method public, but as I mentioned earlier, I'm a little conflicted about it.

Is that use case enough of a reason to expose another public method when I'm gunning for a ruthlessly simple library here? I'm not sure. :upside_down_face:

Reply:
Thoughts, @kyle-rader and @RadicalZephyr?

Reply:
:thinking: 

Okay, how about this paradigm instead. Don't expose the `connect` method, or the slack client (except in the handler functions). If the user wants to connect to slack and do stuff _before_ starting the layabout event loop, let them set it up themselves using the `SlackClient` class directly.  Then, let them pass that same `SlackClient` into their `layabout` instance, either at construction time, or with an attribute setter.  Using a setter would let them still use the decorator style event registration without the slack client needing to be constructed globally too.

That would also have the side benefit that it would give you a way to inject a mock into your test `Layabout` without monkeypatching.

The only downside is that they lose the ability to use your environment variable token retrieval. But that's really just a convenience for making the simple case usage as simple as possible, so it's really not a loss to someone who wants to have more control.